### PR TITLE
Fix tpm2_kdfa for big endian systems

### PR DESCRIPTION
**Pull request by stefanberger**
This patch fixes an issue with tpm2_makeidentity when run on a big
endian system and the result is used by a little endian system, as
may be the case with keylime. The byte swapping must only be done
on little endian systems.

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>

### CODE REVIEW

**Comment by idesai** [comment]
> This patch fixes an issue with tpm2_makeidentity when run on a big
> endian system and the result is used by a little endian system, as
> may be the case with keylime. The byte swapping must only be done
> on little endian systems.
> 
> Signed-off-by: Stefan Berger [stefanb@linux.ibm.com](mailto:stefanb@linux.ibm.com)

@stefanberger since we still have to address #2033, did you happen to test the change on an arm64/32 build?

**Comment by stefanberger** [comment]
> @stefanberger since we still have to address #2033, did you happen to test (run the unit/integrations tests) for the change on an arm64/32 build?

No, a s390 (z system). I don't have access to ARM.

I have  more patches in this direction: https://github.com/stefanberger/tpm2-tools/commits/4.X%2Bbea_fixes

One other issue is that whatever `tpm2_checkquote` seems to look at (from a file) is written in the native endianess of the system that wrote it, so typically little endian today. It's odd to see some of the TPM2 data structures written in little endian, but this is what it is and it's probably least disruptive for the projects to have big endian write the stuff (probably in tpm2_quote) in little endian as well before it's sent from the keylime agent side to the keylime_tenant side that at least looks at it the first time it seems...

File: lib/tpm2_kdfa.c [hunk 1]
```diff
@@ -21,8 +18,7 @@ TSS2_RC tpm2_kdfa(TPMI_ALG_HASH hash_alg, TPM2B *key, char *label,
     tpm2b_i_2.size = 4;
 
     tpm2b_bits.size = 4;
-    bits_swizzled = tpm2_util_endian_swap_32(bits);
-    *(UINT32 *) tpm2b_bits_ptr = bits_swizzled;
+    *(UINT32 *)&tpm2b_bits.buffer[0] = tpm2_util_hton_32(bits);
```
Comment (idesai) [inline lib/tpm2_kdfa.c:21]: Trap representation, makes it concise though.

Comment (stefanberger) [reply lib/tpm2_kdfa.c:21]: What do you mean? 
Does it need a space before the `&`?

Comment (stefanberger) [reply lib/tpm2_kdfa.c:21]: Or do you want me to revert the changes further above and use `tpm2b_bits_ptr` ?

**Comment by idesai** [comment]
- [ ] Test build with qemu-arm32v7

File: lib/tpm2_kdfa.c (hunk 1 above)
Comment (idesai) [reply lib/tpm2_kdfa.c:21]: @stefanberger currently fails arm32v7 build
  CC       lib/lib_libcommon_a-tpm2_kdfa.o
../../lib/tpm2_kdfa.c: In function 'tpm2_kdfa':
../../lib/tpm2_kdfa.c:21:5: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
     *(UINT32 *)&tpm2b_bits.buffer[0] = tpm2_util_hton_32(bits);
     ^
../../lib/tpm2_kdfa.c:57:9: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
         *(UINT32 *)&tpm2b_i_2.buffer[0] = tpm2_util_hton_32(i);
         ^
cc1: all warnings being treated as errors

Comment (idesai) [inline lib/tpm2_kdfa.c:21]:
```suggestion
    UINT8 *tpm2b_buffer_ptr = &tpm2b_bits.buffer[0];
    *(UINT32 *) tpm2b_buffer_ptr  = tpm2_util_hton_32(bits);
```

File: lib/tpm2_kdfa.c [hunk 2]
```diff
@@ -58,9 +54,7 @@ TSS2_RC tpm2_kdfa(TPMI_ALG_HASH hash_alg, TPM2B *key, char *label,
     while (result_key->size < bytes) {
         TPM2B_DIGEST tmpResult;
         // Inner loop
-
-        i_swizzled = tpm2_util_endian_swap_32(i);
-        *(UINT32 *) tpm2b_i_2_ptr = i_swizzled;
+        *(UINT32 *)&tpm2b_i_2.buffer[0] = tpm2_util_hton_32(i);
```
Comment (idesai) [inline lib/tpm2_kdfa.c:57]:
```suggestion
        tpm2b_buffer_ptr = &tpm2b_i_2.buffer[0];
        *(UINT32 *) tpm2b_buffer_ptr  = tpm2_util_hton_32(i);
```

**Review by idesai — CHANGES_REQUESTED** [review]

File: lib/tpm2_kdfa.c (hunk 1 above)
Comment (stefanberger) [reply lib/tpm2_kdfa.c:21]: Your suggestion doesn't work for me (gcc 7.5.0 on s390x):
```
lib/tpm2_kdfa.c: In function ‘tpm2_kdfa’:
lib/tpm2_kdfa.c:22:5: error: invalid operands to binary * (have ‘BYTE * {aka unsigned char *}’ and ‘UINT32 * {aka unsigned int *}’)
     UINT8 *tpm2b_buffer_ptr = &tpm2b_bits.buffer[0]
                               ~~~~~~~~~~~~~~~~~~~~~
     *(UINT32 *) tpm2b_buffer_ptr = tpm2_util_hton_32(bits);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Comment (stefanberger) [reply lib/tpm2_kdfa.c]: Pushed another solution now. Could also fall back to the previous code...

**Comment by williamcroberts** [comment]
For those brave enough, I would also accept a re-write of the KDFA code, that is a nightmare I inherited and never untangled....

File: lib/tpm2_kdfa.c (hunk 1 above)
Comment (stefanberger) [reply lib/tpm2_kdfa.c]: ... with the `tpm2_util_hton_32()` applied of course

Comment (idesai) [reply lib/tpm2_kdfa.c]: > Pushed another solution now. Could also fall back to the previous/original code...

I will post feedback with the build status shortly.

**Comment by idesai** [comment]
> For those brave enough, I would also accept a re-write of the KDFA code, that is a nightmare I inherited and never untangled....

First we will need to add unit/ integration tests including endianness scenarios like these. But I agree, this needs a rewrite.

**Pull request merged**